### PR TITLE
Shor's algorithm, the quantum parts

### PR DESCRIPTION
**Pull request by viathor**
This implements the quantum order finding algorithm.

Example use:

```
$ time python examples/shor.py 21 --order_finder=quantum
7 is a non-trivial factor of 21

real	0m14.425s
user	0m14.758s
sys	0m1.175s
```

Follow-up to #2038.
Fixes https://github.com/quantumlib/Cirq/issues/2134

### CODE REVIEW

File: examples/shor.py [hunk 1]
```diff
@@ -83,9 +95,179 @@ def naive_order_finder(x: int, n: int) -> int:
...
+                a += 1
+            if qubit in self.exponent_register:
+                wire_symbols.append('e' + str(e))
+                e += 1
+        return cirq.CircuitDiagramInfo(wire_symbols=tuple(wire_symbols))
+
+
+def bitlength(n: int) -> int:
+    """Returns minimum number of bits needed to specify a positive integer."""
+    return int(math.ceil(math.log(n) / math.log(2)))
```
Comment (Strilanc) [inline examples/shor.py]: This is a python built-in

```
>>> (4).bit_length()
2
```

File: examples/shor.py [hunk 2]
```diff
@@ -83,9 +95,179 @@ def naive_order_finder(x: int, n: int) -> int:
...
+    Returns:
+        s/r where r is the order of x modulo n and s is in [0..r-1].
+    """
+    exponent_as_integer = result.data['exponent'][0]
+    exponent_num_bits = result.measurements['exponent'].shape[1]
+    return float(exponent_as_integer / 2**exponent_num_bits)
+
+
+def quantum_order_finder(x: int, n: int) -> Optional[int]:
+    """Computes smallest positive r such that x^r mod n == 1.
```
Comment (Strilanc) [inline examples/shor.py]: Use python notation for equations. ** instead of ^

File: examples/shor.py [hunk 3]
```diff
@@ -83,9 +95,179 @@ def naive_order_finder(x: int, n: int) -> int:
...
+        U|y⟩ = |y⟩                n <= y
+
+    in the header of this file.
+
+    Quantum order finding algorithm (which is the quantum part of the Shor's
+    algorithm) uses quantum modular exponentiation together with the Quantum
+    Phase Estimation to compute the order of x modulo n.
+    """
+
+    def __init__(self, ancilla_register: Sequence[cirq.Qid],
```
Comment (Strilanc) [inline examples/shor.py]: Rename to `target`, `exponent`, `base`, `modulus`.

File: examples/shor.py [hunk 4]
```diff
@@ -83,9 +95,179 @@ def naive_order_finder(x: int, n: int) -> int:
...
+        self.n = n
+
+    def registers(self) -> Sequence[Sequence[cirq.Qid]]:
+        return self.ancilla_register, self.exponent_register
+
+    def with_registers(
+            self,
+            *new_registers: Union[int, Sequence['cirq.Qid']],
+    ) -> cirq.ArithmeticOperation:
+        # coverage: ignore
```
Comment (Strilanc) [inline examples/shor.py]: Test this.

File: examples/shor.py [hunk 5]
```diff
@@ -83,9 +95,179 @@ def naive_order_finder(x: int, n: int) -> int:
...
+        assert len(new_registers) == 2
+        assert isinstance(new_registers[0], Sequence)
+        assert isinstance(new_registers[1], Sequence)
```
Comment (Strilanc) [inline examples/shor.py]: The exponent does not technically have to be quantum. It can be classical.

File: examples/shor.py (hunk 2 above)
Comment (viathor) [reply examples/shor.py]: Done throughout.

File: examples/shor.py (hunk 3 above)
Comment (viathor) [reply examples/shor.py]: Much nicer, thanks!

File: examples/shor.py (hunk 4 above)
Comment (viathor) [reply examples/shor.py]: Done (originally left out since it is essentially dead code, unused in this example).

File: examples/shor.py (hunk 5 above)
Comment (viathor) [reply examples/shor.py]: Done.

Initially, I gave this gate the ~minimum functionality necessary for the Shor's algorithm example to work. Now with extra test coverage and additional functionality I wonder whether we should move this gate to cirq.ops. WDYT?

Comment (Strilanc) [reply examples/shor.py]: I don't think it makes sense as an exposed gate, because it is very very *very* non-NISQ.

**Pull request merged**